File: examples/hmac/sign-and-send.mjs
#!/usr/bin/env node
/** Sign and send a private Bitzoom REST request using Node.js built-ins. */

import crypto from 'node:crypto'
import fs from 'node:fs'

function formEncode(value) {
  return encodeURIComponent(value)
    .replace(/[!'()*]/g, (character) =>
      `%${character.charCodeAt(0).toString(16).toUpperCase()}`
    )
    .replace(/%20/g, '+')
}

export function canonicalQuery(rawQuery) {
  const pairs = Array.from(new URLSearchParams(rawQuery).entries())
  const compare = (left, right) => (left < right ? -1 : left > right ? 1 : 0)
  pairs.sort(([leftKey, leftValue], [rightKey, rightValue]) =>
    compare(leftKey, rightKey) || compare(leftValue, rightValue)
  )
  return pairs
    .map(([key, value]) => `${formEncode(key)}=${formEncode(value)}`)
    .join('&')
}

export function sign({
  secret,
  timestamp,
  nonce,
  recvWindow,
  method,
  path,
  rawQuery,
  body,
}) {
  const bodyHash = crypto.createHash('sha256').update(body).digest('hex')
  const payload = [
    timestamp,
    nonce,
    recvWindow,
    method.toUpperCase(),
    path,
    canonicalQuery(rawQuery),
    bodyHash,
  ].join('\n')
  return crypto.createHmac('sha256', secret).update(payload).digest('hex')
}

async function runVectors(vectorPath) {
  const vectors = JSON.parse(fs.readFileSync(vectorPath, 'utf8'))
  const signatures = Object.fromEntries(
    vectors.map((vector) => [
      vector.name,
      sign({ ...vector, body: Buffer.from(vector.body) }),
    ])
  )
  process.stdout.write(JSON.stringify(signatures))
}

async function main() {
  if (process.argv[2] === '--vectors' && process.argv[3]) {
    await runVectors(process.argv[3])
    return
  }

  const apiUrl = (process.env.BITZOOM_HMAC_API_URL || 'https://api1.riverwa.com').replace(/\/+$/, '')
  const apiKey = process.env.BITZOOM_API_KEY
  const apiSecret = process.env.BITZOOM_API_SECRET
  if (!apiKey || !apiSecret) {
    throw new Error('Set BITZOOM_API_KEY and BITZOOM_API_SECRET before running this example.')
  }

  const path = '/api/v1/order'
  const rawQuery = ''
  const body = Buffer.from(
    JSON.stringify({
      symbol: 'BTCUSDT',
      side: 'BUY',
      type: 'LIMIT',
      quantity: '0.001',
      price: '70000',
      clientOrderId: 'bot-1234567890123',
    })
  )
  const timestamp = Date.now().toString()
  const nonce = crypto.randomBytes(16).toString('hex')
  const recvWindow = '5000'
  const signature = sign({
    secret: apiSecret,
    timestamp,
    nonce,
    recvWindow,
    method: 'POST',
    path,
    rawQuery,
    body,
  })
  const response = await fetch(apiUrl + path, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-BZ-APIKEY': apiKey,
      'X-BZ-TIMESTAMP': timestamp,
      'X-BZ-NONCE': nonce,
      'X-BZ-RECVWINDOW': recvWindow,
      'X-BZ-SIGNATURE': signature,
    },
    body,
  })
  const payload = await response.json()
  if (!response.ok || payload.success === false || Number(payload.code || 0) !== 0) {
    throw new Error(payload.errorMessage || `Bitzoom rejected the request: HTTP ${response.status}`)
  }
  console.log(`Request succeeded: HTTP ${response.status}`)
}

await main()

File: examples/hmac/sign_and_send.py
#!/usr/bin/env python3
"""Sign and send a private Bitzoom REST request using only the Python standard library."""

import hashlib
import hmac
import json
import os
import secrets
import sys
import time
from urllib.parse import parse_qsl, quote_plus
from urllib.request import Request, urlopen


def canonical_query(raw_query: str) -> str:
    pairs = parse_qsl(raw_query, keep_blank_values=True)
    pairs.sort(key=lambda pair: (pair[0], pair[1]))
    return "&".join(
        f"{quote_plus(key)}={quote_plus(value)}" for key, value in pairs
    )


def sign(
    secret: str,
    timestamp: str,
    nonce: str,
    recv_window: str,
    method: str,
    path: str,
    raw_query: str,
    body: bytes,
) -> str:
    body_hash = hashlib.sha256(body).hexdigest()
    payload = "\n".join(
        [
            timestamp,
            nonce,
            recv_window,
            method.upper(),
            path,
            canonical_query(raw_query),
            body_hash,
        ]
    ).encode()
    return hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()


def run_vectors(vector_path: str) -> None:
    with open(vector_path, encoding="utf-8") as vector_file:
        vectors = json.load(vector_file)
    signatures = {
        vector["name"]: sign(
            vector["secret"],
            vector["timestamp"],
            vector["nonce"],
            vector["recvWindow"],
            vector["method"],
            vector["path"],
            vector["rawQuery"],
            vector["body"].encode(),
        )
        for vector in vectors
    }
    print(json.dumps(signatures, separators=(",", ":")))


def main() -> None:
    if len(sys.argv) == 3 and sys.argv[1] == "--vectors":
        run_vectors(sys.argv[2])
        return

    api_url = os.environ.get("BITZOOM_HMAC_API_URL", "https://api1.riverwa.com").rstrip("/")
    api_key = os.environ["BITZOOM_API_KEY"]
    api_secret = os.environ["BITZOOM_API_SECRET"]
    path = "/api/v1/order"
    raw_query = ""
    body = json.dumps(
        {
            "symbol": "BTCUSDT",
            "side": "BUY",
            "type": "LIMIT",
            "quantity": "0.001",
            "price": "70000",
            "clientOrderId": "bot-1234567890123",
        },
        separators=(",", ":"),
    ).encode()
    timestamp = str(int(time.time() * 1000))
    nonce = secrets.token_hex(16)
    recv_window = "5000"
    signature = sign(
        api_secret,
        timestamp,
        nonce,
        recv_window,
        "POST",
        path,
        raw_query,
        body,
    )
    request = Request(
        api_url + path,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "X-BZ-APIKEY": api_key,
            "X-BZ-TIMESTAMP": timestamp,
            "X-BZ-NONCE": nonce,
            "X-BZ-RECVWINDOW": recv_window,
            "X-BZ-SIGNATURE": signature,
        },
    )
    with urlopen(request, timeout=15) as response:
        payload = json.loads(response.read())
        if payload.get("success") is False or int(payload.get("code", 0) or 0) != 0:
            raise RuntimeError(payload.get("errorMessage") or "Bitzoom rejected the request")
        print(f"Request succeeded: HTTP {response.status}")


if __name__ == "__main__":
    main()
